library/aes.c

```diff
@@ -497,6 +512,78 @@ static void aes_gen_tables( void )
 
 #endif /* MBEDTLS_AES_ROM_TABLES */
 
+/**
+ * Randomize positions when to use AES SCA countermeasures.
+ * Each byte indicates one AES round as follows:
+ * first ( tbl_len - 2 ) bytes are reserved for AES rounds
+ *  -4 high bit = table to use 0x10 for SCA CM data, 0 otherwise
+ *  -4 low bits = offset based on order, 4 for even position, 0 otherwise
+ * Last 2 bytes for first/final round calculation
+ *  -4 high bit = table to use, 0x10 for SCA CM data, otherwise real data
+ *  -4 low bits = not used
+ *
+ */
+static void aes_sca_cm_data_randomize( uint8_t *tbl, uint8_t tbl_len )
+{
+    int i, is_even_pos;
+#if AES_SCA_CM_ROUNDS != 0
+    int is_unique_number;
+    int num;
+#endif
+
+    memset( tbl, 0, tbl_len );
+
+#if AES_SCA_CM_ROUNDS != 0
+    // Randomize SCA CM positions to tbl
+    for( i = 0; i < AES_SCA_CM_ROUNDS; i++ )
+    {
+        do {
+            is_unique_number = 1;
+/* TODO - Use proper random. This is now ONLY FOR TESTING as mbedtls_platform_random_in_range is alwyays returning 0 */
+            num = /* mbedtls_platform_random_in_range( tbl_len - 1 ) */rand() % (tbl_len - 2);
+
+            if ( tbl[num] == 0 )
+            {
+                is_unique_number = 0;
+                tbl[num] = 0x10;
+            }
+        } while ( is_unique_number == 1 );
+    }
+
+    // Fill start/final round control data
+    num = /* mbedtls_platform_random_in_range( tbl_len - 1 ) */rand() % 0xff;
+    if ( ( num % 2 ) == 0 )
```

Comment:
If I understand correctly, this if-else means that if the first round is "dummy then real" then the final round will be "real then dummy" - and vice versa. Is there a reason for that? Why shouldn't we pick randomly between the 4 possibilities? (DR+DR, DR+RD, RD+DR, RD+RD)

Reply:
No any reason, we can randomise first and last round independently of each other, I'll fix this.